org.jacoco.core/src/org/jacoco/core/internal/analysis/Instruction.java

```diff
@@ -74,14 +73,15 @@
 	public Instruction(final int line) {
 		this.line = line;
 		this.branches = 0;
-		this.coveredBranches = new BitSet();
+		this.coveredBranches = new HashMap<Integer, Integer>();
```

Comment:
Hmm. You might be right. It is hard to see when `Instruction` is used.

At first, I thought `Instruction` was only used during analysis (e.g. report generation). But I see references to `Instruction` from `LabelInfo` as well. In turn, `LabelInfo` is heavily connected to `Label` (ASM construct), which is actively used in instrumentation as well (e.g. even in `ProbeInserter`). In any case, improving this might indeed increase performance by a lot.

Reply:
Implemented using FastUtil's `Int2IntArrayMap`.  HFTC's `MutableMap` has shown do be better performing in large maps. However, we likely don't have large maps for one instruction. Just a few entries for a few branches. Therefore, it is thought that this implementation will be faster.

See for reference: [here](http://java-performance.info/hashmap-overview-jdk-fastutil-goldman-sachs-hppc-koloboke-trove-january-2015/)